fog/ledger/server/tests/router_connection.rs

```diff
@@ -0,0 +1,908 @@
+
+                sleep(Duration::from_secs(10));
+                // panic on the 20th time
+                n += 1; //
+                if n > 20 {
+                    panic!("Fog ledger not  fully initialized");
+                }
+            }
+
+            // FIXME assert_eq!(response.num_txos, ...);
+            assert_eq!(response.results[0].key_image, keys[0]);
+            assert_eq!(response.results[0].status(), Ok(Some(1)));
+            assert_eq!(
+                response.results[0].timestamp_result_code,
+                TimestampResultCode::TimestampFound as u32
+            );
+            assert_eq!(response.results[0].timestamp, 1);
+
+            assert_eq!(response.results[1].key_image, keys[1]);
+            assert_eq!(response.results[1].status(), Ok(Some(1)));
+            assert_eq!(
+                response.results[1].timestamp_result_code,
+                TimestampResultCode::TimestampFound as u32
+            );
+            assert_eq!(response.results[1].timestamp, 1);
+
+            // Check a key_image for a block which will never have signatures & timestamps
+            assert_eq!(response.results[2].key_image, keys[3]);
+            assert_eq!(response.results[2].status(), Ok(Some(2))); // Spent in block 2
+            assert_eq!(
+                response.results[2].timestamp_result_code,
+                TimestampResultCode::TimestampFound as u32
+            );
+            assert_eq!(response.results[2].timestamp, 2);
+
+            // Watcher has only synced 1 block, so timestamp should be behind
+            assert_eq!(response.results[3].key_image, keys[7]);
+            assert_eq!(response.results[3].status(), Ok(Some(3))); // Spent in block 3
+            assert_eq!(
+                response.results[3].timestamp_result_code,
+                TimestampResultCode::TimestampFound as u32
+            );
+            assert_eq!(response.results[3].timestamp, 3);
+
+            // Check a key_image that has not been spent
+            assert_eq!(response.results[4].key_image, keys[19]);
+            assert_eq!(response.results[4].status(), Ok(None)); // Not spent
+            assert_eq!(
+                response.results[4].timestamp_result_code,
+                TimestampResultCode::TimestampFound as u32
+            );
+            assert_eq!(response.results[4].timestamp, u64::MAX);
+        }
+
+        // FIXME: Check a key_image that generates a DatabaseError - tough to generate
+
+        // grpcio detaches all its threads and does not join them :(
+        // we opened a PR here: https://github.com/tikv/grpc-rs/pull/455
+        // in the meantime we can just sleep after grpcio env and all related
+        // objects have been destroyed, and hope that those 6 threads see the
+        // shutdown requests within 1 second.
+        sleep(Duration::from_millis(1000));
+    }
+}
+
+// Test that a fog ledger connection is able to check key images by hitting
+// a fog ledger server
```

Comment:
:question: : this is the same comment as the previous test, how is this different?